File: twigg-web/webcomponents/queue.ts
// Simple queue implementation for O(1) push/pop

export class Entry<T> {
    value: T;
    next: Entry<T> | null = null;
    constructor(value: T) { this.value = value; }
}

export class Fifo<T> {
    private head: Entry<T> | null = null;
    private tail: Entry<T> | null = null;

    // O(1)
    push(value: T) {
        const node = new Entry(value);
        if (!this.tail) {
            this.head = this.tail = node;
        } else {
            this.tail.next = node;
            this.tail = node;
        }
    }

    // O(1)
    pop(): T  {
        if (!this.head){
            throw "popped empty queue"
        };
        const val = this.head.value;
        this.head = this.head.next;
        if (!this.head) this.tail = null;
        return val;
    }

    isEmpty() {
        return this.head === null;
    }

    // pops all entries and returns a slice with them
    drain(): T[] {
        const out: T[] = [];
        while (this.head) out.push(this.pop()!);
        return out;
    }
}

File: base/queue/queue.go
package queue

type node[T any] struct {
	value T
	next  *node[T]
}

type queue[T any] struct {
	head *node[T]
	tail *node[T]
	size int
}

func (q *queue[T]) IsEmpty() bool {
	return q.size == 0
}
func (q *queue[T]) Size() int {
	return q.size
}

func (q *queue[T]) Push(value T) {
	newNode := &node[T]{value: value}

	if q.tail != nil {
		q.tail.next = newNode
	}
	q.tail = newNode

	// If the queue was empty, head should point to the new node
	if q.head == nil {
		q.head = newNode
	}

	q.size++
}

func (q *queue[T]) Pop() T {
	if q.IsEmpty() {
		panic("tried to pop from empty queue")
	}

	value := q.head.value
	q.head = q.head.next
	q.size--

	// If the queue becomes empty, reset the tail to nil
	if q.head == nil {
		q.tail = nil
	}

	return value
}
func (q *queue[T]) Peek() *T {
	if q.IsEmpty() {
		panic("tried to peek at empty queue")
	}
	return &q.head.value
}
